validator/src/bootstrap.rs

```diff
@@ -1295,209 +847,213 @@ mod with_incremental_snapshots {
-        minimal_snapshot_download_speed: f32,
-        maximum_snapshot_download_abort: u64,
-        download_abort_count: &mut u64,
-        snapshot_hash: Option<SnapshotHash>,
-        rpc_contact_info: &ContactInfo,
-    ) -> Result<(), String> {
-        if snapshot_hash.is_none() {
-            return Ok(());
-        }
-        let SnapshotHash {
-            full: full_snapshot_hash,
-            incr: incremental_snapshot_hash,
-        } = snapshot_hash.unwrap();
-
-        // If the local snapshots are new enough, then use 'em; no need to download new snapshots
-        if should_use_local_snapshot(
-            full_snapshot_archives_dir,
-            incremental_snapshot_archives_dir,
-            maximum_local_snapshot_age,
-            full_snapshot_hash,
-            incremental_snapshot_hash,
-        ) {
-            return Ok(());
-        }
-
-        // Check and see if we've already got the full snapshot; if not, download it
-        if snapshot_utils::get_full_snapshot_archives(full_snapshot_archives_dir)
-            .into_iter()
-            .any(|snapshot_archive| {
-                snapshot_archive.slot() == full_snapshot_hash.0
-                    && snapshot_archive.hash() == &full_snapshot_hash.1
-            })
-        {
-            info!(
+    // Check and see if we've already got the full snapshot; if not, download it
+    if snapshot_utils::get_full_snapshot_archives(full_snapshot_archives_dir)
+        .into_iter()
+        .any(|snapshot_archive| {
+            snapshot_archive.slot() == full_snapshot_hash.0
+                && snapshot_archive.hash() == &full_snapshot_hash.1
+        })
+    {
+        info!(
             "Full snapshot archive already exists locally. Skipping download. slot: {}, hash: {}",
             full_snapshot_hash.0, full_snapshot_hash.1
         );
-        } else {
-            download_snapshot(
-                full_snapshot_archives_dir,
-                incremental_snapshot_archives_dir,
-                validator_config,
-                bootstrap_config,
-                use_progress_bar,
-                start_progress,
-                minimal_snapshot_download_speed,
-                maximum_snapshot_download_abort,
-                download_abort_count,
-                rpc_contact_info,
-                full_snapshot_hash,
-                SnapshotType::FullSnapshot,
-            )?;
-        }
+    } else {
+        download_snapshot(
+            full_snapshot_archives_dir,
+            incremental_snapshot_archives_dir,
+            validator_config,
+            bootstrap_config,
+            use_progress_bar,
+            start_progress,
+            minimal_snapshot_download_speed,
+            maximum_snapshot_download_abort,
+            download_abort_count,
+            rpc_contact_info,
+            full_snapshot_hash,
+            SnapshotType::FullSnapshot,
+        )?;
+    }
 
+    if bootstrap_config.incremental_snapshot_fetch {
```

Comment:
I don't think this check needs to be here. The code that calls this function should've ensured that the `SnapshotHash` parameter passed in correctly doesn't have an incremental snapshot component if we're not fetching incremental snapshots.

Optionally, you could add an assert at line 1014 that ensures `snapshot_hash.incr` is `None` if `boostrap_config.incremental_snapshot_fetch` is `false`. (But this assert probably shouldn't be here though; instead likely where `download_snapshots()` is called.)